src/rafimage.cpp

```diff
@@ -248,23 +249,31 @@ void RafImage::readMetadata() {
 
   Internal::enforce(jpg_img_len >= 12, ErrorCode::kerCorruptedMetadata);
 
-  DataBuf buf(jpg_img_len - 12);
-  if (io_->seek(jpg_img_off + 12, BasicIo::beg) != 0)
+  DataBuf jpg_buf(jpg_img_len);
+  if (io_->seek(jpg_img_off, BasicIo::beg) != 0)
     throw Error(ErrorCode::kerFailedToReadImageData);
 
-  if (!buf.empty()) {
-    io_->read(buf.data(), buf.size());
+  if (!jpg_buf.empty()) {
+    io_->read(jpg_buf.data(), jpg_buf.size());
     if (io_->error() || io_->eof())
       throw Error(ErrorCode::kerFailedToReadImageData);
   }
 
-  ByteOrder bo = TiffParser::decode(exifData_, iptcData_, xmpData_, buf.c_data(), buf.size());
+  // Retreive metadata from embedded JPEG preview image.
+  try {
+    auto jpg_io = std::make_unique<Exiv2::MemIo>(jpg_buf.data(), jpg_buf.size());
+    auto jpg_img = JpegImage(std::move(jpg_io), false);
+    jpg_img.readMetadata();
+    setByteOrder(jpg_img.byteOrder());
+    xmpData_ = jpg_img.xmpData();
+    exifData_ = jpg_img.exifData();
+    iptcData_ = jpg_img.iptcData();
+  } catch (const Exiv2::Error&) {
+  }
 
   exifData_["Exif.Image2.JPEGInterchangeFormat"] = getULong(jpg_img_offset, bigEndian);
   exifData_["Exif.Image2.JPEGInterchangeFormatLength"] = getULong(jpg_img_length, bigEndian);
```

Comment:
These should be included in the `try {}` block above.

Reply:
Hm, 50/50 on this one indeed...

Let's leave it as is then, and handle it separately if need be.